x-pack/plugins/apm/server/lib/alerts/register_transaction_duration_alert_type.ts

```diff
@@ -173,7 +176,7 @@ export function registerTransactionDurationAlertType({
           .scheduleActions(alertTypeConfig.defaultActionGroupId, {
             transactionType: alertParams.transactionType,
             serviceName: alertParams.serviceName,
-            environment: environmentParsed.text,
+            environment: getEnvironmentLabel(alertParams.environment),
```

Comment:
Does this still return the correct label when the environment is missing?

Reply:
Hmm looks like the executor implementations are different, I think that's on me, but preferably they all treat environment the same, not sure if that's happening now.

Reply:
What do you mean "missing"?
`getEnvironmentLabel` requires a string. And `alertParams.environment` is not optional.

Reply:
it can be an empty string, which is what we use for the terms agg (`missing: ''`). `alertParams.environment` can probably be `ENVIRONMENT_ALL_VALUE` or `ENVIRONMENT_NOT_DEFINED_VALUE`.

Reply:
> `alertParams.environment` can probably be `ENVIRONMENT_ALL_VALUE` or `ENVIRONMENT_NOT_DEFINED_VALUE`.

This is fine. The implementation will lookup those specific values:


```
const environmentLabels: Record<string, string> = {
  [ENVIRONMENT_ALL_VALUE]: i18n.translate(
    'xpack.apm.filter.environment.allLabel',
    { defaultMessage: 'All' }
  ),
  [ENVIRONMENT_NOT_DEFINED_VALUE]: i18n.translate(
    'xpack.apm.filter.environment.notDefinedLabel',
    { defaultMessage: 'Not defined' }
  ),
};

export function getEnvironmentLabel(environment: string) {
  return environmentLabels[environment] || environment;
}
```

If it's an empty string we have a problem though. I don't think this should happen.

Reply:
There's two sources for environment in rule executors: the alert parameters, which is ENVIRONMENT_ALL_VALUE, ENVIRONMENT_NOT_DEFINED_VALUE, or the specific environment if it's set, and, the value that is returned from the (multi-)terms aggregation, which can be the specific environment if it's set, and an empty string, as we use `missing: ''`. We then use environment to store data in ES (so it should be either undefined or the specific environment), but we also use it as a label in the message, in which case an empty string or ENVIRONMENT_NOT_DEFINED_VALUE should be converted to human-readable text, e.g. "Environment not defined". Does that match your understanding?

Reply:
Hmm.. Instead of `missing: ''` I think we should do: `missing: ENVIRONMENT_NOT_DEFINED.value,` similar to here:

https://github.com/elastic/kibana/blob/bcb941acea6de3067fd515bb1bad5d484cd4b39e/x-pack/plugins/apm/server/lib/environments/get_environments.ts#L64-L71